File: Difficulty: Medium/Top View of Binary Tree/top-view-of-binary-tree.py
# Tree Node
# class Node:
#     def __init__(self, val):
#         self.right = None
#         self.data = val
#         self.left = None

class Solution:
    #Function to return a list of nodes visible from the top view 
    #from left to right in Binary Tree.
    def topView(self,root):
        # code here
        if root is None:
            return []
        
        ans = []
        mpp = {}
        min_level = max_level = 0
        q = deque([(root, 0)])
        
        while q:
            node, vlevel = q.popleft()
            
            if vlevel not in mpp:
                mpp[vlevel] = node.data
            
            # Update min and max levels
            min_level = min(min_level, vlevel)
            max_level = max(max_level, vlevel)
                
            if node.left:
                q.append((node.left, vlevel - 1))
                
            if node.right:
                q.append((node.right, vlevel + 1))
                
        # Iterate from min to max level to avoid sorting
        for level in range(min_level, max_level + 1):
            ans.append(mpp[level])
        
        return ans
                

File: Difficulty: Medium/Top View of Binary Tree/top-view-of-binary-tree.cpp
/*
struct Node
{
    int data;
    Node* left;
    Node* right;
};
*/
class Solution {
  public:
    // Function to return a list of nodes visible from the top view
    // from left to right in Binary Tree.
    vector<int> topView(Node *root) {
        // code here
        vector<int> ans;
        if(root==NULL) return ans;
        map<int,int> mpp;
        queue<pair<Node*,int>> q;
        q.push({root,0});
        while(!q.empty()){
            auto it = q.front();
            q.pop();
            Node* node = it.first;
            int line = it.second;
            if(mpp.find(line)==mpp.end()) mpp[line]= node->data;
            
            if(node->left !=NULL){
                q.push({node->left,line-1});
            }
            if(node->right !=NULL){
                q.push({node->right,line+1});
            }
        }
        for(auto it : mpp){
            ans.push_back(it.second);
        }
        return ans;
    }
};